pkg/local-pv-monitor/monitor.go

```diff
@@ -0,0 +1,508 @@
+			mountPath, err = common.GetContainerPath(pv, config)
+			if err != nil {
+				glog.Errorf("get container path error: %v", err)
+			}
+			break
+		}
+	}
+	if len(mountPath) == 0 {
+		// can not find mount path, this may because: admin modify config(hostpath)
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, HostPathNotExist, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	dir, dirErr := monitor.VolUtil.IsDir(mountPath)
+	bl, blErr := monitor.VolUtil.IsBlock(mountPath)
+	if !dir && !bl && (dirErr != nil || blErr != nil) {
+		// mountPath does not exist or is not a directory
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, HostPathNotExist, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	continueThisCheck = true
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, HostPathNotExist, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+
+}
+
+func (monitor *LocalPVMonitor) checkPVAndFSSize(mountPath string, pv *v1.PersistentVolume) {
+	capacityByte, err := monitor.VolUtil.GetFsCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q fs stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two values must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than FS capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
+
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+
+}
+
+func (monitor *LocalPVMonitor) checkPVAndBlockSize(mountPath string, pv *v1.PersistentVolume) {
+	capacityByte, err := monitor.VolUtil.GetBlockCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q block stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two values must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than block device capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
```

Comment:
yeah, we should do this.

updating PV with capacity will pose unreasonable pressure on API server, to surface this, i'm thinking of dong threshold based approach. for example, we pv status is CapacityPressure if 80% of its capacity has been used.

Reply:
>updating PV with capacity will pose unreasonable pressure on API server, to surface this, i'm thinking of dong threshold based approach. for example, we pv status is CapacityPressure if 80% of its capacity has been used.

sounds good, can do this at next stage